drag_and_drop_v2/public/js/drag_and_drop.js

```diff
@@ -1,388 +1,325 @@
+            state.state.feedback = data.feedback;
+            if (data.finished) {
+                state.state.finished = true;
+                state.feedback.finish = data.final_feedback;
+            }
+            setState(state);
+        });
+    };
+
+    var closePopup = function(evt) {
+        var target = $(evt.target);
+        var popup_box = '.xblock--drag-and-drop .popup';
+        var close_button = '.xblock--drag-and-drop .popup .close';
+        var submit_input_button = '.xblock--drag-and-drop .submit-input';
+        var state = getState();
+
+        if (!state.state.feedback) {
+            return;
+        }
+        if (target.is(popup_box) || target.is(submit_input_button)) {
+            return;
+        }
+        if (target.parents(popup_box).length && !target.is(close_button)) {
+            return;
+        }
+
+        publish_event({
+            event_type: 'xblock.drag-and-drop-v2.feedback.closed',
+            content: state.state.feedback,
+            manually: true
+        });
+
+        delete state.state.feedback;
+        setState(state);
+    };
+
+    var resetExercise = function() {
+        $.ajax({
+            type: 'POST',
+            url: runtime.handlerUrl(element, 'reset'),
+            data: '{}',
+            success: setState
+        });
+    };
+
+    var px = function(n) {
+        return n + 'px';
+    };
+
+    var renderItems = function(data) {
+        return data.items.map(function(item) {
+            var item_state = data.state.items[item.id];
+            var position = item_state || {};
+            var input = null;
+            if (item.inputOptions) {
+                var has_input = Boolean(item_state && 'input' in item_state);
+                var className;
+                if (has_input && !item_state.submitting_input) {
+                    className = item_state.correct_input ? 'correct' : 'incorrect';
                 }
-            },
-
-            data: null
-        };
-
-        return {
-            init: _fn.init,
-        };
-    })(jQuery);
-
-    $.ajax(runtime.handlerUrl(element, 'get_data'), {
-        dataType: 'json'
-    }).done(function(data){
-        dragAndDrop.init(data);
-    });
-
-    publish_event({event_type:"xblock.drag-and-drop-v2.loaded"});
+                var is_input_visible = item_state && !item_state.submitting_location;
+                input = (
+                    h('div.numerical-input', {className: className,
+                                              style: {display: is_input_visible ? 'block' : 'none'}},
+                      [
+                          h('input.input', {type: 'text', disabled: has_input,
+                                            focusHook: has_input ? undefined : FocusHook(),
+                                            value: has_input ? item_state.input : ''}),
+                          h('button.submit-input', {disabled: has_input}, 'ok')
```

Comment:
I like the virtual-dom refactor, but one drawback is that you are now including HTML in the JS code. I imagine it isn't really possible to use templating with this approach? If not, can we at least keep remain as close to the way we would approach this via templating - ie extracting the definition to separate functions outside of the current scope, in a different JS file, and passing variables explicitly via a context object like you did below with `data`? To try to minimize the temptation to intertwine code and presentation, and keep these separate as much as possible.